gtsam/sfm/MFAS.cpp

```diff
@@ -0,0 +1,138 @@
+
+using namespace gtsam;
+using std::pair;
+using std::vector;
+
+MFAS::MFAS(const std::shared_ptr<vector<Key>> &nodes,
+           const std::shared_ptr<TranslationEdges> &relativeTranslations,
+           const Unit3 &projection_direction)
+    : nodes_(nodes), relativeTranslations_(relativeTranslations),
+      relativeTranslationsForWeights_(std::make_shared<TranslationEdges>()) {
+  // iterate over edges and flip all edges that have negative weights,
+  // while storing the magnitude of the weights.
+  for (auto it = relativeTranslations->begin();
+       it != relativeTranslations->end(); it++) {
+    KeyPair edge = it->first;
+    double weight = it->second.dot(projection_direction);
+    if (weight < 0.0) {
+      std::swap(edge.first, edge.second);
+      weight *= -1;
+    }
+    positiveEdgeWeights_[edge] = weight;
+  }
+}
+
+MFAS::MFAS(const std::shared_ptr<std::vector<Key>> &nodes,
+           const std::map<KeyPair, double> &edgeWeights) : nodes_(nodes),
+                                                           relativeTranslations_(std::make_shared<TranslationEdges>()),
+                                                           relativeTranslationsForWeights_(std::make_shared<
+                                                               TranslationEdges>()) {
+  // similar to the above direction constructor, but here weights are
+  // provided as input.
+  for (auto it = edgeWeights.begin(); it != edgeWeights.end(); it++) {
+    KeyPair edge = it->first;
+
+    // When constructed like this, we do not have access to the relative translations. 
+    // So, we store the unswapped edge in the relativeTranslationsForWeights_ map with a default 
+    // Unit3 value. This helps retain the original direction of the edge in the returned result
+    // of computeOutlierWeights
+    relativeTranslationsForWeights_->insert({edge, Unit3()});
+
+    double weight = it->second;
+    if (weight < 0.0) {
+      // change the direction of the edge to make weight positive
+      std::swap(edge.first, edge.second);
+      weight *= -1;
+    }
+    positiveEdgeWeights_[edge] = weight;
+  }
+}
+
+std::vector<Key> MFAS::computeOrdering() {
+  FastMap<Key, double> in_weights;    // sum on weights of incoming edges for a node
+  FastMap<Key, double> out_weights;   // sum on weights of outgoing edges for a node
+  FastMap<Key, vector<Key> > in_neighbors;
+  FastMap<Key, vector<Key> > out_neighbors;
+
+  // populate neighbors and weights
+  for (auto it = positiveEdgeWeights_.begin(); it != positiveEdgeWeights_.end(); it++) {
+    const KeyPair &edge = it->first;
+    in_weights[edge.second] += it->second;
+    out_weights[edge.first] += it->second;
+    in_neighbors[edge.second].push_back(edge.first);
+    out_neighbors[edge.first].push_back(edge.second);
+  }
+
+  // in each iteration, one node is appended to the ordered list
+  while (orderedNodes_.size() < nodes_->size()) {
+
+    // finding the node with the max heuristic score
```

Comment:
separate this in a function?

Reply:
It is not used anywhere else, but yes, improves readability to factor that into a different function. I shall make that change in the next iteration if no other changes are suggested.